rdmd.d

```diff
@@ -916,3 +891,71 @@ void yap(size_t line = __LINE__, T...)(auto ref T stuff)
     debug stderr.writeln(line, ": ", stuff);
     else stderr.writeln(stuff);
 }
+
+/**
+Used to wrap filesystem operations that should also be logged with the
+yap function or affected by dryRun. Append the string "IfLive" to the end
+of the function for it to be skipped during a dry run.
+
+These functions allow the filename to be given once so the log statement
+always matches the operation. Using it also guarantees you won't forget to
+include a `yap` alongside any file operation you want to be logged.
+*/
+struct Filesystem
+{
+static:
+    static auto opDispatch(string func, T...)(T args)
+    {
+        static if (func.endsWith("IfLive"))
+        {
+            enum fileFunc = func[0 .. $ - "IfLive".length];
+            enum skipOnDryRun = true;
+        }
+        else
+        {
+            enum fileFunc = func;
+            enum skipOnDryRun = false;
+        }
+
+        static if ([fileFunc].findAmong(["exists", "timeLastModified", "isFile", "existsAsFile"]))
+            yap("stat ", args[0]);
+        else static if (fileFunc == "rename")
+            yap("mv ", args[0], " ", args[1]);
+        else static if ([fileFunc].findAmong(["remove", "mkdirRecurse", "rmdirRecurse", "dirEntries", "write", "touch"]))
+            yap(fileFunc, " ", args[0]);
+        else static assert(0, "Filesystem.opDispath has not implemented " ~ fileFunc);
+
+        static if (skipOnDryRun)
+        {
+            if (dryRun)
+                return;
+        }
+        mixin("return DirectFilesystem." ~ fileFunc ~ "(args);");
+    }
+}
+
+/**
+Operates on the file system without logging its operations or being
+affected by dryRun.
+*/
+struct DirectFilesystem
+{
+static:
+    import file = std.file;
+    alias file this;
+
+    /**
+    Update a file's timestamp.
+    */
+    static void touch(string name)
+    {
+        file.write(name, "");
+    }
+    /**
+    Returns true if name exists and is a file.
+    */
+    static bool existsAsFile(string name)
+    {
+        return file.exists(name) && file.isFile(name);
+    }
```

Comment:
The purpose of this method still isn't really clear.  If the point of the chatty output is to allow a point of comparison with `strace` output (for example), then this just obscures things: while only one `stat` will be output by `yap`, two `stat` calls will be reported by `strace`.

Consider e.g. the following "rdmd lite" for comparison:

```D
void main (string[] args)
{
    import std.file : exists, isFile;
    if (args[1].exists && args[1].isFile)
    {
        import std.stdio : writefln;
        writefln("'%s' exists!", args[1]);
    }
}
```

Now, run this:

```
$ strace ./rdmd_lite hello.d 2>&1 | grep -e "hello\.d"
execve("./rdmd_lite", ["./rdmd_lite", "hello.d"], [/* 72 vars */]) = 0
lstat("hello.d", {st_mode=S_IFREG|0664, st_size=140, ...}) = 0
stat("hello.d", {st_mode=S_IFREG|0664, st_size=140, ...}) = 0
write(1, "'hello.d' exists!\n", 18'hello.d' exists!
```

Reply:
> If the point of the chatty output is to allow a point of comparison with strace output

That's not the point.  Look at the comment above the `Filesytem` struct and read it...

Reply:
`std.file` is not meant to have 1:1 mapping to OS syscalls, so I think this isn't something to worry about. Anyhow I understood this can be optimized to just one syscall in a later change.

Reply:
> `std.file` is not meant to have 1:1 mapping to OS syscalls

No, but we have at least one case in `rdmd.d` that goes:

```D
if (Filesystem.exists(...))
{
    enforce(Filesystem.isFile(...), ...);
}
```

which means that in that one case, we have a different behaviour from the cases that use `existsAsFile`.  We can eliminate this discrepancy by eliminating `existsAsFile`.

In the process, that also saves us implementing a new method definition, so where's the downside in doing so?

Reply:
Sounds to me that `existsAsFile` is the wrong primitive here. We want a function that returns three discernable values:

1. Exists and is file
2. Exists but is not a file
3. Does not exist.

I recall the proposed implementation for `existsAsFile` used `getAttributes` to a similar effect. Maybe that would be a more correct direction?

Reply:
This is yet another repetition of the same arguments.  This has already been addressed, I think no answer is necessary.

Reply:
> Sounds to me that existsAsFile is the wrong primitive here. We want a function that returns three discernable values:

I think `existsAsFile` works well in the places it's used.  For this one case, using another function such as `getAtributes` would work well, but I'll put that in a separate PR.

Reply:
But why bother?

What is being asked here is that you do not introduce an EXTRA method that does not need to exist.  There is no obvious downside to the non-existence of that method.  Ergo, _there is no reason to introduce it_.

Reply:
@CyberShadow:

> We want a function that returns three discernable values

Why do we _need_ any new function, rather than just using the existing wrapped support for `exists` and `isFile`?

Reply:
I've already explained this multiple times. I don't see how explaining it one more time is going to help.

Reply:
What you've done is to present `existsAsFile` as a fait-accompli and then reject my arguments for _not_ having it.

But it's not a reviewer's responsibility to argue the case to _drop_ a proposed new function.  It's the code author's responsibility to make a _positive_ case that it is genuinely beneficial to have it.

You haven't made that case, whereas there are several reasons why _having_ it introduces some unnecessary inconsistencies in app behaviour.

Reply:
> Why do we need any new function, 

See https://github.com/dlang/tools/pull/306 - but never mind (see my comment there).

Reply:
> What you've done is to present existsAsFile as a fait-accompli and then reject my arguments for not having it.
>
> But it's not a reviewer's responsibility to argue the case to drop a proposed new function. It's the code author's responsibility to make a positive case that it is genuinely beneficial to have it.

Talking with you is like talking to a goldfish...you just keep repeating yourself and forget what I'v already said.

The reason for it is to reduce noise in the log that doesn't help understand what rdmd is doing.  That's it...it's that simple.  Please don't ask me to justify this again.